diff --git a/samples/CMakeLists.txt b/samples/CMakeLists.txt
index c1c5ae699..4dcfeec1a 100644
--- a/samples/CMakeLists.txt
+++ b/samples/CMakeLists.txt
@@ -68,7 +68,7 @@ endif()
 # Automatically include the multiprotocol_rpmsg as child image when both RPMSG HCI
 # and RPMSG 802.15.4 Serialization are enabled or the hci_rpmsg sample when
 # only RPMSG HCI is enabled. For any of these samples change the board to be the network core.
-if (CONFIG_BT_RPMSG_NRF53)
+if (CONFIG_BT_RPMSG_NRF53 OR CONFIG_BLE_NETWORK_CORE)
   if (CONFIG_SOC_NRF5340_CPUAPP)
 
     if (CONFIG_NRF_802154_SER_HOST)
@@ -78,9 +78,15 @@ if (CONFIG_BT_RPMSG_NRF53)
       "CONFIG_BT_RPMSG_NRF53 and CONFIG_NRF_802154_SER_HOST are set to 'y'")
     else()
       set(NETCORE_IMAGE "hci_rpmsg")
-      set(NETCORE_IMAGE_PATH "${ZEPHYR_BASE}/samples/bluetooth/${NETCORE_IMAGE}")
-      message("Adding 'hci_rpmsg' firmware as child image since "
-      "CONFIG_BT_RPMSG_NRF53 is set to 'y'")
+      if (CONFIG_BLE_NETWORK_CORE)
+        set(NETCORE_IMAGE_PATH "${ZEPHYR_BASE}/../nrf53_ble/ble_netcore")
+        message("Adding 'ble_netcore' firmware as child image since "
+        "CONFIG_BLE_NETWORK_CORE is set to 'y'")	  
+      else()
+        set(NETCORE_IMAGE_PATH "${ZEPHYR_BASE}/samples/bluetooth/${NETCORE_IMAGE}")
+        message("Adding 'hci_rpmsg' firmware as child image since "
+        "CONFIG_BT_RPMSG_NRF53 is set to 'y'")
+      endif()
     endif()
 
     if (CONFIG_BT_LL_SOFTDEVICE_DEFAULT)
