diff --git a/boot/bootutil/src/loader.c b/boot/bootutil/src/loader.c
index 716e3ed..0c3f036 100644
--- a/boot/bootutil/src/loader.c
+++ b/boot/bootutil/src/loader.c
@@ -44,6 +44,8 @@
 #include "bootutil/security_cnt.h"
 #include "bootutil/boot_record.h"
 #include "bootutil/fault_injection_hardening.h"
+#include <storage/flash_map.h>
+#include <drivers/flash.h>
 
 #ifdef CONFIG_SOC_NRF5340_CPUAPP
 #include <dfu/pcd.h> 
@@ -729,14 +731,37 @@ boot_validated_swap_type(struct boot_loader_state *state,
     int swap_type;
     fih_int fih_rc = FIH_FAILURE;
     bool upgrade_valid = false;
-
+    
 #if defined(PM_S1_ADDRESS) || defined(CONFIG_SOC_NRF5340_CPUAPP)
-    const struct flash_area *secondary_fa =
-        BOOT_IMG_AREA(state, BOOT_SECONDARY_SLOT);
-    struct image_header *hdr = (struct image_header *)secondary_fa->fa_off;
+    const struct flash_area *secondary_fa;
+    struct image_header *hdr;
     uint32_t vtable_addr = 0;
     uint32_t *vtable = 0;
     uint32_t reset_addr = 0;
+    bool secondary_in_internal_flash;
+
+    secondary_fa = BOOT_IMG_AREA(state, BOOT_SECONDARY_SLOT);
+
+     if (strcmp(secondary_fa->fa_dev_name, "NRF_FLASH_DRV_NAME") == 0)
+    {
+        printk("=secondary slot in internal Flash=");
+        secondary_in_internal_flash = true;
+    }
+    else
+    {
+        printk("= secondary slot in External Flash %x =", secondary_fa->fa_off);
+        secondary_in_internal_flash = false;
+    }   
+    
+    if (secondary_in_internal_flash)
+    {
+        hdr = (struct image_header *)secondary_fa->fa_off;
+    }
+    else
+    {
+        hdr = boot_img_hdr(state, BOOT_SECONDARY_SLOT);
+    }
+   
     /* Patch needed for NCS. Since image 0 (the app) and image 1 (the other
      * B1 slot S0 or S1) share the same secondary slot, we need to check
      * whether the update candidate in the secondary slot is intended for
@@ -746,9 +771,23 @@ boot_validated_swap_type(struct boot_loader_state *state,
      */
 
     if (hdr->ih_magic == IMAGE_MAGIC) {
-        vtable_addr = (uint32_t)hdr + hdr->ih_hdr_size;
-        vtable = (uint32_t *)(vtable_addr);
-        reset_addr = vtable[1];
+        if (secondary_in_internal_flash)
+        {        
+            vtable_addr = (uint32_t)hdr + hdr->ih_hdr_size;
+            vtable = (uint32_t *)(vtable_addr);
+            reset_addr = vtable[1];
+            printk(" In vtable %x, reset addr %x ", vtable_addr, reset_addr); 
+        }
+        else
+        {                       
+            vtable_addr = hdr->ih_hdr_size;
+            int rc = flash_area_read(secondary_fa, vtable_addr+4, &reset_addr, sizeof(reset_addr));
+            if (rc != 0) {
+                return BOOT_SWAP_TYPE_FAIL;
+            }
+            vtable = (uint32_t *) (0x10000000 + vtable_addr); //0x10000000 is XIP base address            
+            printk(" Ext vtable %x, reset addr %x ", (uint32_t)vtable, reset_addr);             
+        }
 #ifdef PM_S1_ADDRESS
         const struct flash_area *primary_fa;
         int rc = flash_area_open(flash_area_id_from_multi_image_slot(
@@ -793,16 +832,53 @@ boot_validated_swap_type(struct boot_loader_state *state,
          */
         if (upgrade_valid && reset_addr > PM_CPUNET_B0N_ADDRESS) {
             uint32_t fw_size = hdr->ih_img_size;
-
+            // if (!secondary_in_internal_flash)
+            // {
+            //     BOOT_LOG_INF("Enable XIP feature");
+            //     NRF_QSPI_S->XIPEN = 1;  //enable XIP feature
+            // }
             BOOT_LOG_INF("Starting network core update");
             int rc = pcd_network_core_update(vtable, fw_size);
+            // if (!secondary_in_internal_flash)
+            // {
+            //     NRF_QSPI_S->ENABLE = 0;
+            //     printk("disable and enable QSPI");                                
+            //     NRF_QSPI_S->ENABLE = 1;                
+            // }            
 
             if (rc != 0) {
                 swap_type = BOOT_SWAP_TYPE_FAIL;
             } else {
                 BOOT_LOG_INF("Done updating network core");
-                rc = swap_erase_trailer_sectors(state,
-                        secondary_fa);
+                if (secondary_in_internal_flash)
+                {
+                    rc = swap_erase_trailer_sectors(state,
+                            secondary_fa);
+                    if (rc != 0) {
+                        printk("1error with erasing image trailor");
+                    }        
+                }
+                else
+                {
+                    //This is a workaround, since qspi nor flash fail to set the write_protection
+                    uint32_t off;
+                    uint32_t sz;
+                    uint32_t sector;
+                    const struct device *flash_dev;
+
+                    sector = boot_img_num_sectors(state, BOOT_SECONDARY_SLOT) - 1;
+                    sz = boot_img_sector_size(state, BOOT_SECONDARY_SLOT, sector);
+                    off = boot_img_sector_off(state, BOOT_SECONDARY_SLOT, sector);
+                    printk("**sz=%x,off=%x, areaSize=%x**", sz, off, secondary_fa->fa_size);
+                    if (off <= secondary_fa->fa_size && off + sz <= secondary_fa->fa_size)        
+                    {
+                        flash_dev = device_get_binding(secondary_fa->fa_dev_name);
+                        rc = flash_erase(flash_dev, secondary_fa->fa_off + off, sz);
+                        if (rc != 0) {
+                            printk("2error with erasing image trailor");
+                        } 
+                    }                  
+                }
                 swap_type = BOOT_SWAP_TYPE_NONE;
             }
         }
@@ -824,7 +900,7 @@ boot_validated_swap_type(struct boot_loader_state *state,
  */
 int
 boot_erase_region(const struct flash_area *fap, uint32_t off, uint32_t sz)
-{
+{    
     return flash_area_erase(fap, off, sz);
 }
 
diff --git a/boot/bootutil/src/swap_misc.c b/boot/bootutil/src/swap_misc.c
index 940d646..25ae0d2 100644
--- a/boot/bootutil/src/swap_misc.c
+++ b/boot/bootutil/src/swap_misc.c
@@ -46,14 +46,14 @@ swap_erase_trailer_sectors(const struct boot_loader_state *state,
     uint8_t image_index;
     int rc;
 
-    BOOT_LOG_DBG("erasing trailer; fa_id=%d", fap->fa_id);
-
+    BOOT_LOG_INF("erasing trailer; fa_id=%d dev_name=%s", fap->fa_id, fap->fa_dev_name);
+    printk("fa size=%x, offset=%x ", fap->fa_size, fap->fa_off);
     image_index = BOOT_CURR_IMG(state);
     fa_id_primary = flash_area_id_from_multi_image_slot(image_index,
             BOOT_PRIMARY_SLOT);
     fa_id_secondary = flash_area_id_from_multi_image_slot(image_index,
             BOOT_SECONDARY_SLOT);
-
+    printk("fap id =%d fa_id_primary =%d fa_id_secondary=%d", fap->fa_id, fa_id_primary, fa_id_secondary);
     if (fap->fa_id == fa_id_primary) {
         slot = BOOT_PRIMARY_SLOT;
     } else if (fap->fa_id == fa_id_secondary) {
@@ -68,12 +68,13 @@ swap_erase_trailer_sectors(const struct boot_loader_state *state,
     total_sz = 0;
     do {
         sz = boot_img_sector_size(state, slot, sector);
-        off = boot_img_sector_off(state, slot, sector);
+        off = boot_img_sector_off(state, slot, sector);        
         rc = boot_erase_region(fap, off, sz);
         assert(rc == 0);
 
         sector--;
         total_sz += sz;
+        printk("sz=%d total_sz=%d trailer_sz=%d off=%x", sz, total_sz, trailer_sz, off);
     } while (total_sz < trailer_sz);
 
     return rc;
